zebra-chain/src/history_tree.rs

```diff
@@ -0,0 +1,245 @@
+//! History tree (Merkle mountain range) structure that contains information about
+//! the block history as specified in ZIP-221.
+
+use std::{
+    collections::{BTreeMap, HashSet},
+    io,
+    sync::Arc,
+};
+
+use thiserror::Error;
+
+use crate::{
+    block::{Block, ChainHistoryMmrRootHash, Height},
+    orchard,
+    parameters::{Network, NetworkUpgrade},
+    primitives::zcash_history::{Entry, Tree},
+    sapling,
+};
+
+/// An error describing why a history tree operation failed.
+#[derive(Debug, Error)]
+#[non_exhaustive]
+#[allow(missing_docs)]
+pub enum HistoryTreeError {
+    #[error("zcash_history error: {inner:?}")]
+    #[non_exhaustive]
+    InnerError { inner: zcash_history::Error },
+
+    #[error("I/O error")]
+    IOError(#[from] io::Error),
+}
+
+/// History tree structure.
```

Comment:
We can steal some docs 👼

```suggestion
/// History tree (Merkle mountain range) structure that contains information about
// the block history, as specified in [ZIP-221][https://zips.z.cash/zip-0221].
```